crates/block-producer/src/block_builder/mod.rs

```diff
@@ -130,47 +255,79 @@ impl BlockBuilder {
             .filter(|note_id| !output_notes_set.contains(note_id))
             .collect();
 
-        // Request information needed for block building from the store
-        let block_inputs = self
-            .store
-            .get_block_inputs(
-                updated_account_set.into_iter(),
-                produced_nullifiers.iter(),
-                dangling_notes.iter(),
-            )
-            .await
-            .map_err(BuildBlockError::GetBlockInputsFailed)?;
-
-        let missing_notes: Vec<_> = dangling_notes
-            .difference(&block_inputs.found_unauthenticated_notes.note_ids())
-            .copied()
-            .collect();
-        if !missing_notes.is_empty() {
-            return Err(BuildBlockError::UnauthenticatedNotesNotFound(missing_notes));
+        Self {
+            updated_accounts,
+            nullifiers,
+            output_notes,
+            dangling_notes,
         }
+    }
+}
 
-        let (block_header_witness, updated_accounts) = BlockWitness::new(block_inputs, batches)?;
+struct SelectedBlock {
+    block_number: BlockNumber,
+    batches: Vec<ProvenBatch>,
+}
+struct BlockSummaryAndInputs {
+    batches: Vec<ProvenBatch>,
+    summary: BlockSummary,
+    inputs: BlockInputs,
+}
+struct ProvenBlock {
+    block: Block,
+}
 
-        let new_block_header = self.block_kernel.prove(block_header_witness)?;
+impl SelectedBlock {
+    fn inject_telemetry(&self) {
+        let span = Span::current();
+        span.set_attribute("block.number", i64::from(self.block_number.as_u32()));
+        span.set_attribute("block.batches.count", i64::from(self.batches.len() as u32));
+    }
+}
 
-        // TODO: return an error?
-        let block =
-            Block::new(new_block_header, updated_accounts, output_notes, produced_nullifiers)
-                .expect("invalid block components");
+impl BlockSummaryAndInputs {
+    fn inject_telemetry(&self) {
+        let span = Span::current();
 
-        let block_hash = block.hash();
-        let block_num = new_block_header.block_num();
+        // SAFETY: We do not expect to have more than u32::MAX of any count per block.
+        span.set_attribute(
+            "block.updated_accounts.count",
+            i64::try_from(self.summary.updated_accounts.len())
+                .expect("less than u32::MAX account updates"),
+        );
+        span.set_attribute(
+            "block.output_notes.count",
+            i64::try_from(self.summary.output_notes.len())
+                .expect("less than u32::MAX output notes"),
+        );
```

Comment:
`BlockSummary::output_notes` is a `Vec<Vec<OutputNote>>`, so we have to sum up the number of output notes of each inner vec.

Reply:
Ah right. Thank you!